### PR TITLE
Templatize cmake and bakefile builds

### PR DESCRIPTION
**Pull request by randombit**
Merge of #1343 #1344 #1345

Will rebase before merge to clean up the history.

### CODE REVIEW

**Comment by randombit** [comment]
OK this is ready to go now.

**Comment by neusdan** [comment]
Bakefile x64 build works fine. x86 doesn't work:

`python configure.py --with-bakefile --cpu=x86`

Output from Bakefile:

`botan.bkl:488:8: error: variable "archs" (list of architectures): expression "x86_32" is not a valid architecture value: must be one of "x86", "x86_64"`

**Comment by randombit** [comment]
Ah ok I'll fix that, thanks for testing.

File: configure.py [hunk 1]
```diff
@@ -1114,6 +1122,30 @@ def isa_flags_for(self, isa, arch):
...
+
+                if isa is None:
+                    continue
+
+            flagset = self.isa_flags_for(isa, arch.basename)
+            if flagset is None:
+                raise UserError('Compiler %s does not support %s' % (self.basename, isa))
+            flags.add(flagset)
+
+        return ("".join([" %s" % f for f in sorted(flags)])).strip()
```
Comment (webmaster128) [inline configure.py]: Isn't this the same as `" ".join(sorted(flags))`?

Comment (randombit) [reply configure.py]: I believe so, will change. Thanks

**Comment by randombit** [comment]
@neusdan Fixed the bakefile arch issue, also I noticed in the bakefile docs that it accepts having multiple files listed in the `sources {}` block so I changed the output a bit in a299f08. Should be identical behavior? Let me know if that ends up causing problems for you.

**Comment by randombit** [comment]
Will merge this tomorrow night, assuming no additional problems are identified.

**Comment by neusdan** [comment]
Tested it again. Everything fine!

**Pull request merged**